src/GraphBLAS-sharp.Backend/Vector/COOVector/COOVector.fs

```diff
@@ -0,0 +1,542 @@
+                   && allIndices.[gid] = allIndices.[gid + 1] then
+                    positions.[gid] <- 0
+
+                    match (%opAdd) (Both(leftValues.[gid], rightValues.[gid + 1])) with
+                    | Some value ->
+                        allValues.[gid + 1] <- value
+                        positions.[gid + 1] <- 1
+                    | None -> positions.[gid + 1] <- 0
+                elif (gid < length
+                      && gid > 0
+                      && allIndices.[gid - 1] <> allIndices.[gid])
+                     || gid = 0 then
+                    if isLeft.[gid] = 1 then
+                        match (%opAdd) (Left(leftValues.[gid])) with
+                        | Some value ->
+                            allValues.[gid] <- value
+                            positions.[gid] <- 1
+                        | None -> positions.[gid] <- 0
+                    else
+                        match (%opAdd) (Right(rightValues.[gid])) with
+                        | Some value ->
+                            allValues.[gid] <- value
+                            positions.[gid] <- 1
+                        | None -> positions.[gid] <- 0 @>
+
+        let kernel = clContext.Compile(preparePositions)
+
+        fun (processor: MailboxProcessor<_>) (allIndices: ClArray<int>) (leftValues: ClArray<'a>) (rightValues: ClArray<'b>) (isLeft: ClArray<int>) ->
+
+            let length = allIndices.Length
+
+            let allValues =
+                clContext.CreateClArray(
+                    length,
+                    hostAccessMode = HostAccessMode.NotAccessible,
+                    deviceAccessMode = DeviceAccessMode.ReadWrite,
+                    allocationMode = AllocationMode.Default
+                )
+
+            let positions =
+                clContext.CreateClArray(
+                    length,
+                    hostAccessMode = HostAccessMode.NotAccessible,
+                    deviceAccessMode = DeviceAccessMode.ReadWrite,
+                    allocationMode = AllocationMode.Default
+                )
+
+            let ndRange =
+                Range1D.CreateValid(length, workGroupSize)
+
+            let kernel = kernel.GetKernel()
+
+            processor.Post(
+                Msg.MsgSetArguments
+                    (fun () ->
+                        kernel.KernelFunc ndRange length allIndices leftValues rightValues isLeft allValues positions)
+            )
+
+            processor.Post(Msg.CreateRunMsg<_, _>(kernel))
+
+            allValues, positions
+
+    let private setPositions<'a when 'a: struct> (clContext: ClContext) (workGroupSize: int) =
+
+        let setPositions =
+            <@ fun (ndRange: Range1D) prefixSumArrayLength (allValues: ClArray<'a>) (allIndices: ClArray<int>) (prefixSumBuffer: ClArray<int>) (resultValues: ClArray<'a>) (resultIndices: ClArray<int>) ->
+
+                let i = ndRange.GlobalID0
+
+                if i = prefixSumArrayLength - 1
```

Comment:
This confition may cause errors in rare cases. Fixed version is [here](https://github.com/YaccConstructor/GraphBLAS-sharp/blob/835e7bfee1507ee5c31f5613e175368e7daf00d7/src/GraphBLAS-sharp.Backend/Matrix/COOMatrix/COOMatrix.fs#L13)